pkg/storegateway/indexheader/index/postings.go

```diff
@@ -0,0 +1,514 @@
+				}
+				valueIndex++
+				if valueIndex == len(values) {
+					break
+				}
+				wantedValue = values[valueIndex]
+			}
+
+			if i+1 == len(e.offsets) {
+				// No more offsets for this name.
+				// Break this loop and record lastOffset on the way for ranges we just added if any.
+				for j := range newSameRngs {
+					newSameRngs[j].End = e.lastValOffset
+				}
+				rngs = append(rngs, newSameRngs...)
+				break
+			}
+
+			if valueIndex != len(values) && wantedValue <= e.offsets[i+1].value {
+				// wantedValue is smaller or same as the next offset we know about, let's iterate further to add those.
+				continue
+			}
+
+			// Nothing wanted or wantedValue is larger than next offset we know about.
+			// Let's exit and do binary search again / exit if nothing wanted.
+
+			if len(newSameRngs) > 0 {
+				// We added some ranges in this iteration. Use next posting offset as the end of our ranges.
+				// We know it exists as we never go further in this loop than e.offsets[i, i+1].
+
+				skipNAndName(&d, &buf)
+				d.UvarintBytes() // Label value.
+				postingOffset := int64(d.Uvarint64())
+
+				for j := range newSameRngs {
+					newSameRngs[j].End = postingOffset - crc32.Size
+				}
+				rngs = append(rngs, newSameRngs...)
+			}
+			break
+		}
+		if d.Err() != nil {
+			return nil, errors.Wrap(d.Err(), "get postings offset entry")
+		}
+	}
+
+	return rngs, nil
+}
+
+func (t *PostingOffsetTableV2) LabelValues(name string, filter func(string) bool) ([]string, error) {
+	e, ok := t.postings[name]
+	if !ok {
+		return nil, nil
+	}
+	if len(e.offsets) == 0 {
+		return nil, nil
+	}
+	values := make([]string, 0, len(e.offsets)*t.postingOffsetsInMemSampling)
+
+	// Don't Crc32 the entire postings offset table, this is very slow
+	// so hope any issues were caught at startup.
+	d := t.factory.NewDecbufAtUnchecked(t.tableOffset)
+	defer t.factory.Close(d)
+
+	d.Skip(e.offsets[0].tableOff)
+	lastVal := e.offsets[len(e.offsets)-1].value
+
+	skip := 0
+	for d.Err() == nil {
+		if skip == 0 {
+			// These are always the same number of bytes,
+			// and it's faster to skip than parse.
+			skip = d.Len()
+			d.Uvarint()      // Keycount.
+			d.UvarintBytes() // Label name.
+			skip -= d.Len()
+		} else {
+			d.Skip(skip)
+		}
+		s := yoloString(d.UvarintBytes()) // Label value.
+		if filter == nil || filter(s) {
+			// Clone the yolo string since its bytes will be invalidated as soon as
+			// any other reads against the decoding buffer are performed.
+			values = append(values, strings.Clone(s))
+		}
+		if s == lastVal {
+			break
+		}
+		d.Uvarint64() // Offset.
+	}
+	if d.Err() != nil {
+		return nil, errors.Wrap(d.Err(), "get postings offset entry")
+	}
+	return values, nil
+}
+
+func (t *PostingOffsetTableV2) LabelNames() ([]string, error) {
+	labelNames := make([]string, 0, t.LabelNameCount())
+
+	err := t.ForEachLabelName(func(name string) error {
+		labelNames = append(labelNames, name)
+		return nil
+	})
```

Comment:
I think we can make this much more efficiently without reusing this `ForEachLabelName` call by just copying the code iterating `t.postings` from below.